tooling/nargo_fmt/src/formatter.rs

```diff
@@ -0,0 +1,303 @@
+    }
+
+    fn write_right_bracket(&mut self) {
+        self.write_token(Token::RightBracket);
+    }
+
+    fn write_comma(&mut self) {
+        self.write_token(Token::Comma);
+    }
+
+    fn write_semicolon(&mut self) {
+        self.write_token(Token::Semicolon);
+    }
+
+    /// Writes the given keyword, if the current token is that keyword
+    /// (so this is a check that we are producing a token we expect to be in the source
+    /// we are traversing). Then advances to the next token.
+    ///
+    /// Calls `write_token` so comments and whitespaces are skipped before writing the keyword.
+    fn write_keyword(&mut self, keyword: Keyword) {
+        self.write_token(Token::Keyword(keyword));
+    }
+
+    /// Writes the given token, if the current token is the same as the given one
+    /// (so this is a check that we are producing a token we expect to be in the source
+    /// we are traversing). Then advances to the next token.
+    ///
+    /// Before writing the token any comments and spaces are skipped. This is so that
+    /// a caller can call `write_token`, `write_keyword`, `write_space`, etc., without
+    /// having to explicitly call `skip_comments_and_whitespace` in between those calls.
+    fn write_token(&mut self, token: Token) {
+        self.skip_comments_and_whitespace();
+        if self.token == token {
+            self.write_current_token();
+            self.bump();
+        } else {
+            panic!("Expected token {:?}, got: {:?}", token, self.token);
+        }
+    }
+
+    /// Writes the current token but doesn't advance to the next one.
+    fn write_current_token(&mut self) {
+        self.write(&self.token.to_string());
+    }
+
+    /// Writes the current token trimming its end but doesn't advance to the next one.
+    /// Mainly used when writing comment lines, because we never want trailing spaces
+    /// inside comments.
+    fn write_current_token_trimming_end(&mut self) {
+        self.write(self.token.to_string().trim_end());
+    }
+
+    /// Writes the current token but without turning it into a string using `to_string()`.
+    /// Instead, we check the token's span and format what's in the original source there
+    /// (useful when formatting integer tokens, because a token like 0xFF ends up being an
+    /// integer with a value 255, but we don't want to change 0xFF to 255).
+    fn write_current_token_as_in_source(&mut self) {
+        self.write_source_span(self.token_span);
+    }
+
+    /// Writes whatever is in the given span relative to the file's source that's being formatted.
+    fn write_source_span(&mut self, span: Span) {
+        self.write(&self.source[span.start() as usize..span.end() as usize]);
+    }
+
+    /// Writes the current indentation to the buffer, but only if the buffer
+    /// does not end with a space (otherwise we'd be indenting too much).
+    fn write_indentation(&mut self) {
+        if self.buffer.ends_with(' ') {
```

Comment:
I don't fully understand why only space is checked. Shouldn't indentation go only after a new line?

Reply:
Good point! I think this ended up being called twice in some cases (because of nested calls where we can't know if some nested formatting ended up writing a final indentation) so just checking that it doesn't end with a space was enough. But your idea is better (I just tried it and nothing breaks :-)). I'll push a commit soon to fix this.